File: java/src/test/java/com/bakobo/fiki/UnitTest.java
package com.bakobo.fiki;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

/**
 * The ground the shared vectors do not cover: signing a fresh request, generating a key, the
 * parser's own refusal branches, and the freshness rules beyond the three cases refusals.json
 * pins. Those are not cross-implementation contracts — they are this port working.
 */
class UnitTest {

    private static final String SEED_AID = "BAOhB7_zzhC-HXDdGOdLwJln5NYwm6UNXx3chmQSVTG4";
    private static final String URL_QUERY = "https://api.example.com/things?limit=1&sort=name";
    private static final long SIGNED_AT = 1_700_000_000L;
    private static final byte[] BODY = "{\"hello\": \"world\"}".getBytes(StandardCharsets.UTF_8);

    private static Key key() {
        byte[] seed = new byte[32];
        for (int i = 0; i < 32; i++) {
            seed[i] = (byte) i;
        }
        return Key.fromSeed(seed);
    }

    private static Map<String, String> headers(String... pairs) {
        Map<String, String> out = new LinkedHashMap<>();
        for (int i = 0; i < pairs.length; i += 2) {
            out.put(pairs[i], pairs[i + 1]);
        }
        return out;
    }

    private static String line(String component, String method, String url, Map<String, String> hdrs) {
        byte[] base = Fiki.signatureBase(method, url, hdrs, List.of(component),
            Fiki.Params.of(SIGNED_AT, "k"));
        return new String(base, StandardCharsets.UTF_8).split("\n")[0];
    }

    private static Map<String, String> signed(Fiki.SignOptions opts) {
        Fiki.SignOptions withCreated = opts.created() == null ? opts.withCreated(SIGNED_AT) : opts;
        return Fiki.signRequest(key(), "POST", URL_QUERY, Map.of(), withCreated);
    }

    private static FikiException.Kind kindOf(Executable body) {
        return assertThrows(FikiException.class, body::run).kind();
    }

    private interface Executable {
        void run();
    }

    @Test
    void theSeedYieldsTheAidEveryOtherPortDerives() {
        assertEquals(SEED_AID, key().aid());
        assertEquals(32, key().seed().length);
        assertNotEquals(key().aid(), Key.generate().aid());
        assertEquals(SEED_AID, Key.toAid(Key.verifyingKeyBytes(SEED_AID)));
    }

    @Test
    void aSeedOfTheWrongLengthIsRefused() {
        assertEquals(FikiException.Kind.MalformedKey, kindOf(() -> Key.fromSeed(new byte[31])));
        assertEquals(FikiException.Kind.MalformedKey, kindOf(() -> Key.fromSeed(null)));
    }

    @ParameterizedTest
    @ValueSource(strings = {
        "BAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",   // too short
        "BAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA", // too long
        "DAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",  // transferable prefix
        "B!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!",  // outside the alphabet
        // "=" is inside base64's alphabet, so a lenient decoder would take this and decode short.
        "BAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
    })
    void malformedAidsAreRefused(String aid) {
        assertEquals(FikiException.Kind.MalformedKey, kindOf(() -> Key.verifyingKeyBytes(aid)));
    }

    @Test
    void derivedComponents() {
        assertEquals("\"@authority\": api.example.com",
            line("@authority", "GET", "/things", headers("Host", "API.example.com")));
        assertEquals("\"@authority\": example.com",
            line("@authority", "GET", "https://EXAMPLE.com:443/f", Map.of()));
        assertEquals("\"@authority\": example.com:8443",
            line("@authority", "GET", "https://example.com:8443/f", Map.of()));
        assertEquals("\"@path\": /", line("@path", "GET", "https://example.com", Map.of()));
        assertEquals("\"@path\": /f", line("@path", "GET", "https://example.com/f#frag", Map.of()));
        assertEquals("\"@query\": ?", line("@query", "GET", "https://example.com/f", Map.of()));
        assertEquals("\"@query\": ?baz=bat%2Dman",
            line("@query", "GET", "https://example.com/p?baz=bat%2Dman", Map.of()));
        assertEquals("\"@method\": POST", line("@method", "post", "https://example.com/f", Map.of()));
        assertEquals("\"content-type\": application/json",
            line("Content-Type", "GET", "https://x.example/f", headers("Content-Type", "  application/json  ")));
    }

    @Test
    void unbuildableAndMissingComponentsAreRefused() {
        assertEquals(FikiException.Kind.UnsupportedComponent,
            kindOf(() -> line("@target-uri", "GET", "https://x.example/f", Map.of())));
        assertEquals(FikiException.Kind.MissingComponent,
            kindOf(() -> line("x-absent", "GET", "https://x.example/f", Map.of())));
        assertEquals(FikiException.Kind.MissingComponent,
            kindOf(() -> line("@authority", "GET", "/things", Map.of())));
    }

    @Test
    void optionalParametersSerializeInAFixedOrder() {
        byte[] base = Fiki.signatureBase("GET", "https://x.example/f", Map.of(), List.of("@method"),
            new Fiki.Params(SIGNED_AT, "k", "ed25519", SIGNED_AT + 60, "abc", "app"));
        String[] lines = new String(base, StandardCharsets.UTF_8).split("\n");
        assertEquals(
            "\"@signature-params\": (\"@method\");created=1700000000;expires=1700000060;"
                + "nonce=\"abc\";alg=\"ed25519\";keyid=\"k\";tag=\"app\"",
            lines[lines.length - 1]);
    }

    @Test
    void signAndVerifyRoundTrip() {
        Map<String, String> out = signed(Fiki.SignOptions.none().withBody(BODY));
        assertTrue(out.containsKey("Content-Digest"));
        Fiki.Verdict verdict = Fiki.verifyRequest("POST", URL_QUERY, out,
            Fiki.VerifyOptions.decliningFreshness().withBody(BODY));
        assertEquals(SEED_AID, verdict.aid());
        assertTrue(verdict.covered().contains("content-digest"));
    }

    @Test
    void theKeyidCarriesTheRawKeyRatherThanTheAid() {
        Map<String, String> out = signed(Fiki.SignOptions.none());
        String keyid = out.get("Signature-Input").split("keyid=\"")[1].split("\"")[0];
        assertEquals(43, keyid.length());
    }

    @Test
    void aChosenCoveredSetOmittingTheDigestRefusesABody() {
        assertEquals(FikiException.Kind.UncoveredBody, kindOf(() -> Fiki.signRequest(
            key(), "POST", URL_QUERY, Map.of(),
            Fiki.SignOptions.none().withBody(BODY).withCovered(List.of("@method")))));
    }

    @Test
    void aChosenCoveredSetIncludingTheDigestSignsABody() {
        Map<String, String> out = signed(Fiki.SignOptions.none().withBody(BODY)
            .withCovered(List.of("@method", "@path", "content-digest")).withLabel("mine"));
        assertTrue(out.get("Signature-Input").startsWith("mine="));
        Fiki.verifyRequest("POST", URL_QUERY, out,
            Fiki.VerifyOptions.decliningFreshness().withBody(BODY));
    }

    @Test
    void aCallerSuppliedDigestIsUsedRatherThanRecomputed() {
        Map<String, String> supplied = headers("Content-Digest", Fiki.contentDigest(BODY));
        Map<String, String> out = Fiki.signRequest(key(), "POST", URL_QUERY, supplied,
            Fiki.SignOptions.none().withBody(BODY).withCreated(SIGNED_AT));
        assertFalse(out.containsKey("Content-Digest"), "fiki should not echo back a digest it was given");
        Map<String, String> all = new LinkedHashMap<>(supplied);
        all.putAll(out);
        Fiki.verifyRequest("POST", URL_QUERY, all,
            Fiki.VerifyOptions.decliningFreshness().withBody(BODY));
    }

    @Test
    void signingWithoutACreatedUsesTheWallClock() {
        Map<String, String> out =
            Fiki.signRequest(key(), "GET", URL_QUERY, null, Fiki.SignOptions.none());
        Fiki.verifyRequest("GET", URL_QUERY, out, Fiki.VerifyOptions.maxAge(300));
    }

    @Test
    void anExpectedAidIsAuthoritativeOverTheInlineKeyid() {
        Map<String, String> out = signed(Fiki.SignOptions.none());
        assertEquals(SEED_AID, Fiki.verifyRequest("POST", URL_QUERY, out,
            Fiki.VerifyOptions.decliningFreshness().withExpectedAid(SEED_AID)).aid());
        String stranger = Key.generate().aid();
        assertEquals(FikiException.Kind.SignatureMismatch, kindOf(() -> Fiki.verifyRequest(
            "POST", URL_QUERY, out, Fiki.VerifyOptions.decliningFreshness().withExpectedAid(stranger))));
    }

    @Test
    void aMalformedExpectedAidIsRefused() {
        Map<String, String> out = signed(Fiki.SignOptions.none());
        assertEquals(FikiException.Kind.MalformedKey, kindOf(() -> Fiki.verifyRequest(
            "POST", URL_QUERY, out, Fiki.VerifyOptions.decliningFreshness().withExpectedAid("nope"))));
    }

    @Test
    void aCoveredComponentTheVerifierCannotBuildIsRefused() {
        Map<String, String> out = new LinkedHashMap<>(signed(Fiki.SignOptions.none()));
        out.put("Signature-Input",
            out.get("Signature-Input").replaceFirst("\\(\"@method\"", "(\"@target-uri\""));
        assertEquals(FikiException.Kind.UnsupportedComponent, kindOf(() ->
            Fiki.verifyRequest("POST", URL_QUERY, out, Fiki.VerifyOptions.decliningFreshness())));
    }

    @Test
    void digestHandling() {
        record Case(String digest, FikiException.Kind expected) {}
        List<Case> cases = List.of(
            new Case("sha-1=:AAAA:, " + Fiki.contentDigest(BODY), null),
            new Case("sha-1=:AAAA:", FikiException.Kind.MalformedDigest),
            new Case("sha-256=\"not bytes\"", FikiException.Kind.MalformedDigest),
            new Case("((( not sfv", FikiException.Kind.MalformedDigest));
        for (Case c : cases) {
            Map<String, String> supplied = headers("Content-Digest", c.digest());
            Map<String, String> out = Fiki.signRequest(key(), "POST", URL_QUERY, supplied,
                Fiki.SignOptions.none().withBody(BODY).withCreated(SIGNED_AT));
            Map<String, String> all = new LinkedHashMap<>(supplied);
            all.putAll(out);
            if (c.expected() == null) {
                Fiki.verifyRequest("POST", URL_QUERY, all,
                    Fiki.VerifyOptions.decliningFreshness().withBody(BODY));
            } else {
                assertEquals(c.expected(), kindOf(() -> Fiki.verifyRequest("POST", URL_QUERY, all,
                    Fiki.VerifyOptions.decliningFreshness().withBody(BODY))), c.digest());
            }
        }
    }

    @Test
    void freshness() {
        Map<String, String> out = signed(Fiki.SignOptions.none());
        Fiki.verifyRequest("POST", URL_QUERY, out, Fiki.VerifyOptions.maxAge(300).withNow(SIGNED_AT + 299));
        Fiki.verifyRequest("POST", URL_QUERY, out, Fiki.VerifyOptions.maxAge(300).withNow(SIGNED_AT + 303));
        assertEquals(FikiException.Kind.SignatureTooOld, kindOf(() -> Fiki.verifyRequest(
            "POST", URL_QUERY, out, Fiki.VerifyOptions.maxAge(300).withNow(SIGNED_AT + 400))));
        assertEquals(FikiException.Kind.SignatureTooOld, kindOf(() -> Fiki.verifyRequest(
            "POST", URL_QUERY, out, Fiki.VerifyOptions.maxAge(300).withSkew(0).withNow(SIGNED_AT + 301))));
        assertEquals(FikiException.Kind.SignatureTooOld, kindOf(() -> Fiki.verifyRequest(
            "POST", URL_QUERY, out, Fiki.VerifyOptions.maxAge(300).withNow(SIGNED_AT - 60))));
        Fiki.verifyRequest("POST", URL_QUERY, out,
            Fiki.VerifyOptions.decliningFreshness().withNow(SIGNED_AT + 1_000_000));
    }

    @Test
    void expiresIsEnforcedEvenWhenMaxAgeIsDeclined() {
        Map<String, String> out = signed(Fiki.SignOptions.none().withExpires(SIGNED_AT + 60));
        Fiki.verifyRequest("POST", URL_QUERY, out,
            Fiki.VerifyOptions.decliningFreshness().withNow(SIGNED_AT + 30));
        assertEquals(FikiException.Kind.SignatureExpired, kindOf(() -> Fiki.verifyRequest(
            "POST", URL_QUERY, out, Fiki.VerifyOptions.decliningFreshness().withNow(SIGNED_AT + 66))));
    }

    @ParameterizedTest
    @ValueSource(strings = {
        "1=2", "a=\"oops", "a=\"o\\ps\"", "a=:not base64!:", "a=:AAAA", "a=?2", "a=?",
        "a=%bad", "a=(\"@method\"", "a=(\"@method\";q=1)", "a=(1)", "a=(\"@method\"\"@path\")",
        "a=1 b=2", "a=1, ", "a=-", "a=", "a;x=%", "a=(%)",
    })
    void theParserRefusesWhatItShould(String text) {
        assertThrows(Sfv.SyntaxException.class, () -> Sfv.parseDictionary(text));
    }

    @Test
    void theParserRoundTripsWhatRfc9421PutsInSignatureInput() {
        List<Sfv.Member> parsed =
            Sfv.parseDictionary("sig=(\"@method\" \"@path\");created=1;keyid=\"k\";alg=\"ed25519\"");
        assertEquals("(\"@method\" \"@path\");created=1;keyid=\"k\";alg=\"ed25519\"",
            Sfv.serializeInnerList(parsed.get(0).list()));
        assertEquals("(\"a\\\"b\\\\c\");f;g=?0", Sfv.serializeInnerList(new Sfv.InnerList(
            List.of("a\"b\\c"), List.of(Map.entry("f", Boolean.TRUE), Map.entry("g", Boolean.FALSE)))));
    }

    @Test
    void theParserReadsTheShapesRfc8941AllowsHere() {
        for (String text : List.of("a=?1", "a=?0", "a=-12", "a", "a;x", "a=()", "  a=1  ", "a=:AAAA:")) {
            Sfv.parseDictionary(text);
        }
    }
}

File: js/test/unit.test.js
// The ground the shared vectors do not cover.
//
// The vectors pin what every port must agree on: bases, signatures, refusal classes. They say
// nothing about signing a fresh request, generating a key, or the parser's own error branches,
// because those are not cross-implementation contracts — they are this port working. So they are
// tested here, at the same 100% branch gate the Python port holds.

import assert from 'node:assert/strict';
import { describe, it } from 'node:test';

import {
  DEFAULT_COVERED,
  Key,
  contentDigest,
  errors,
  signRequest,
  signatureBase,
  toAid,
  verifyRequest,
  verifyingKey,
} from '../src/index.js';
import { parseDictionary, serializeInnerList } from '../src/sfv.js';

const SEED = Uint8Array.from({ length: 32 }, (_, i) => i);
const AID = 'BAOhB7_zzhC-HXDdGOdLwJln5NYwm6UNXx3chmQSVTG4';
const URL_WITH_QUERY = 'https://api.example.com/things?limit=1&sort=name';
const BODY = new TextEncoder().encode('{"hello": "world"}');
const SIGNED_AT = 1700000000;

const key = await Key.fromSeed(SEED);

async function signed(overrides = {}) {
  const args = { key, method: 'POST', url: URL_WITH_QUERY, headers: {}, body: BODY, created: SIGNED_AT, ...overrides };
  const headers = { ...(args.headers ?? {}), ...(await signRequest(args)) };
  return { request: { method: args.method, url: args.url, body: args.body }, headers };
}

describe('keys', () => {
  it('derives the AID keripy produces for the same seed', () => {
    assert.equal(key.aid, AID);
  });

  it('generates a distinct key each time, non-extractable by default', async () => {
    const a = await Key.generate();
    const b = await Key.generate();
    assert.notEqual(a.aid, b.aid);
    assert.equal(a.aid.length, 44);
    assert.throws(() => a.seed, errors.MalformedKey);
  });

  it('hands back a seed only when the caller asked for a portable identity', async () => {
    const portable = await Key.generate({ extractable: true });
    assert.equal(portable.seed.length, 32);
    assert.equal((await Key.fromSeed(portable.seed)).aid, portable.aid);
  });

  it('keeps the seed of an extractable fromSeed key', async () => {
    const portable = await Key.fromSeed(SEED, { extractable: true });
    assert.deepEqual(portable.seed, SEED);
  });

  it('refuses a seed of the wrong length', async () => {
    await assert.rejects(() => Key.fromSeed(new Uint8Array(31)), errors.MalformedKey);
    await assert.rejects(() => Key.fromSeed('not bytes'), errors.MalformedKey);
  });

  for (const [id, aid] of [
    ['too short', 'B' + 'A'.repeat(42)],
    ['too long', 'B' + 'A'.repeat(44)],
    ['transferable prefix', 'D' + 'A'.repeat(43)],
    ['not a string', 42],
    // "=" is inside base64's alphabet, so a 44-character string of the right shape still decodes
    // short. A decoder is exactly the place a quiet shortfall turns into somebody else's error.
    ['padded, decodes short', 'B' + 'A'.repeat(41) + '=='],
    ['outside the alphabet', 'B' + '!'.repeat(43)],
  ]) {
    it(`refuses an AID that is ${id}`, () => {
      assert.throws(() => verifyingKey(aid), errors.MalformedKey);
    });
  }

  it('round-trips a raw key through the lens', () => {
    assert.equal(toAid(verifyingKey(AID)), AID);
  });
});

describe('the signature base', () => {
  const line = (component, overrides = {}) =>
    new TextDecoder()
      .decode(
        signatureBase({
          method: 'POST',
          url: URL_WITH_QUERY,
          headers: {},
          covered: [component],
          created: SIGNED_AT,
          keyid: 'k',
          ...overrides,
        }),
      )
      .split('\n')[0];

  it('takes the authority from the Host header when the URL is relative', () => {
    assert.equal(line('@authority', { url: '/things', headers: { Host: 'API.example.com' } }), '"@authority": api.example.com');
  });

  it('refuses @authority with neither a URL authority nor a Host header', () => {
    assert.throws(() => line('@authority', { url: '/things' }), errors.MissingComponent);
  });

  it('omits a default port and keeps a non-default one', () => {
    assert.equal(line('@authority', { url: 'https://EXAMPLE.com:443/f' }), '"@authority": example.com');
    assert.equal(line('@authority', { url: 'https://example.com:8443/f' }), '"@authority": example.com:8443');
  });

  it('treats an empty path as the slash the origin server sees', () => {
    assert.equal(line('@path', { url: 'https://example.com' }), '"@path": /');
  });

  it('binds "no query" as a bare question mark', () => {
    assert.equal(line('@query', { url: 'https://example.com/f' }), '"@query": ?');
  });

  it('uppercases the method and lowercases header names', () => {
    assert.equal(line('@method', { method: 'post' }), '"@method": POST');
    assert.equal(line('Content-Type', { headers: { 'Content-Type': '  application/json  ' } }), '"content-type": application/json');
  });

  it('refuses a derived component it cannot build, rather than skipping it', () => {
    assert.throws(() => line('@target-uri'), errors.UnsupportedComponent);
  });

  it('refuses a covered header the request lacks', () => {
    assert.throws(() => line('x-absent'), errors.MissingComponent);
  });

  it('serializes the optional parameters in a fixed order', () => {
    const base = new TextDecoder().decode(
      signatureBase({
        method: 'POST', url: 'https://example.com/f', headers: {}, covered: ['@method'],
        created: SIGNED_AT, keyid: 'k', alg: 'ed25519', expires: SIGNED_AT + 60, nonce: 'abc', tag: 'app',
      }),
    );
    assert.equal(
      base.split('\n').at(-1),
      `"@signature-params": ("@method");created=${SIGNED_AT};expires=${SIGNED_AT + 60};nonce="abc";alg="ed25519";keyid="k";tag="app"`,
    );
  });
});

describe('signing and verifying', () => {
  it('round-trips and names the signer', async () => {
    const { request, headers } = await signed();
    const verdict = await verifyRequest({ ...request, headers, maxAge: null });
    assert.equal(verdict.aid, AID);
    for (const component of DEFAULT_COVERED) assert.ok(verdict.covered.includes(component));
  });

  it('carries the raw key in keyid rather than the AID', async () => {
    const { headers } = await signed();
    const keyid = headers['Signature-Input'].split('keyid="')[1].split('"')[0];
    assert.equal(keyid.length, 43);
  });

  it('digests a body and covers the digest', async () => {
    const { request, headers } = await signed();
    assert.ok(headers['Content-Digest']);
    const verdict = await verifyRequest({ ...request, headers, maxAge: null });
    assert.ok(verdict.covered.includes('content-digest'));
  });

  it('uses a caller-supplied Content-Digest rather than recomputing it', async () => {
    const supplied = { 'Content-Digest': await contentDigest(BODY) };
    const { request, headers } = await signed({ headers: supplied });
    assert.equal(headers['Content-Digest'], supplied['Content-Digest']);
    assert.equal((await verifyRequest({ ...request, headers, maxAge: null })).aid, AID);
  });

  it('refuses to sign a body under a chosen covered set that omits the digest', async () => {
    await assert.rejects(
      () => signRequest({ key, method: 'POST', url: URL_WITH_QUERY, headers: {}, body: BODY, covered: ['@method'] }),
      errors.UncoveredBody,
    );
  });

  it('signs a body under a chosen covered set that includes the digest', async () => {
    const { request, headers } = await signed({ covered: ['@method', '@path', 'content-digest'] });
    assert.equal((await verifyRequest({ ...request, headers, maxAge: null })).aid, AID);
  });

  it('needs no digest for a bodyless request', async () => {
    const { request, headers } = await signed({ body: null });
    assert.equal(headers['Content-Digest'], undefined);
    assert.equal((await verifyRequest({ ...request, headers, maxAge: null })).aid, AID);
  });

  it('uses the wall clock when no created is given', async () => {
    const { request, headers } = await signed({ created: null });
    assert.equal((await verifyRequest({ ...request, headers, maxAge: 300 })).aid, AID);
  });

  it('treats an expected AID as authoritative over the inline keyid', async () => {
    const { request, headers } = await signed();
    assert.equal((await verifyRequest({ ...request, headers, expectedAid: AID, maxAge: null })).aid, AID);
    const stranger = (await Key.fromSeed(Uint8Array.from({ length: 32 }, (_, i) => i + 1))).aid;
    await assert.rejects(
      () => verifyRequest({ ...request, headers, expectedAid: stranger, maxAge: null }),
      errors.SignatureMismatch,
    );
  });

  it('refuses a request with no freshness policy stated at all', async () => {
    const { request, headers } = await signed();
    await assert.rejects(() => verifyRequest({ ...request, headers }), TypeError);
  });

  it('refuses a digest naming only algorithms it cannot compute', async () => {
    const { request, headers } = await signed({ headers: { 'Content-Digest': 'sha-1=:AAAA:' } });
    await assert.rejects(() => verifyRequest({ ...request, headers, maxAge: null }), errors.MalformedDigest);
  });

  it('accepts a digest naming an unknown algorithm alongside one it knows', async () => {
    const supplied = { 'Content-Digest': `sha-1=:AAAA:, ${await contentDigest(BODY)}` };
    const { request, headers } = await signed({ headers: supplied });
    assert.equal((await verifyRequest({ ...request, headers, maxAge: null })).aid, AID);
  });
});

describe('freshness', () => {
  it('accepts a signature inside max age and refuses one outside it', async () => {
    const { request, headers } = await signed();
    assert.equal((await verifyRequest({ ...request, headers, maxAge: 300, now: SIGNED_AT + 299 })).aid, AID);
    await assert.rejects(
      () => verifyRequest({ ...request, headers, maxAge: 300, now: SIGNED_AT + 400 }),
      errors.SignatureTooOld,
    );
  });

  it('tolerates skew, and lets the allowance be tightened', async () => {
    const { request, headers } = await signed();
    assert.equal((await verifyRequest({ ...request, headers, maxAge: 300, now: SIGNED_AT + 303 })).aid, AID);
    await assert.rejects(
      () => verifyRequest({ ...request, headers, maxAge: 300, skew: 0, now: SIGNED_AT + 301 }),
      errors.SignatureTooOld,
    );
  });

  it('refuses a created in the future beyond the skew allowance', async () => {
    const { request, headers } = await signed();
    await assert.rejects(
      () => verifyRequest({ ...request, headers, maxAge: 300, now: SIGNED_AT - 60 }),
      errors.SignatureTooOld,
    );
  });

  it('enforces the signer\'s own expires even when max age is declined', async () => {
    const { request, headers } = await signed({ expires: SIGNED_AT + 60 });
    assert.equal((await verifyRequest({ ...request, headers, maxAge: null, now: SIGNED_AT + 30 })).aid, AID);
    await assert.rejects(
      () => verifyRequest({ ...request, headers, maxAge: null, now: SIGNED_AT + 66 }),
      errors.SignatureExpired,
    );
  });

  it('refuses a max-age check against a signature carrying no created', async () => {
    // RFC 9421 makes created optional, so a foreign signer may omit it. fiki's own never does,
    // and the signature has to be genuinely made this way — freshness is checked after the
    // signature, so a doctored Signature-Input just fails the signature instead.
    const covered = ['@method', '@path'];
    const inner = { items: covered, params: new Map([['keyid', key.keyid]]) };
    const { componentLines } = await import('../src/base.js');
    const lines = componentLines({ method: 'GET', url: '/a', headers: {}, covered });
    lines.push(`"@signature-params": ${serializeInnerList(inner)}`);
    const signature = await key.sign(new TextEncoder().encode(lines.join('\n')));
    const headers = {
      'Signature-Input': `sig=${serializeInnerList(inner)}`,
      Signature: `sig=:${Buffer.from(signature).toString('base64')}:`,
    };
    assert.equal((await verifyRequest({ method: 'GET', url: '/a', headers, maxAge: null })).aid, AID);
    await assert.rejects(
      () => verifyRequest({ method: 'GET', url: '/a', headers, maxAge: 300, now: SIGNED_AT }),
      errors.SignatureTooOld,
    );
  });
});

describe('the structured-fields subset', () => {
  it('round-trips what RFC 9421 puts in Signature-Input', () => {
    const parsed = parseDictionary('sig=("@method" "@path");created=1;keyid="k";alg="ed25519"');
    assert.equal(
      serializeInnerList(parsed.get('sig')),
      '("@method" "@path");created=1;keyid="k";alg="ed25519"',
    );
  });

  it('reads the shapes RFC 8941 allows here', () => {
    assert.equal(parseDictionary('a=?1, b=?0').get('a').value, true);
    assert.equal(parseDictionary('a=?0').get('a').value, false);
    assert.equal(parseDictionary('a=-12').get('a').value, -12);
    assert.equal(parseDictionary('a').get('a').value, true);
    assert.equal(parseDictionary('a;x').get('a').params.get('x'), true);
    assert.equal(parseDictionary('a="say \\"hi\\" \\\\"').get('a').value, 'say "hi" \\');
    assert.deepEqual([...parseDictionary('a=()').get('a').items], []);
    assert.equal(parseDictionary('  a=1  ').get('a').value, 1);
  });

  it('escapes on the way back out', () => {
    assert.equal(
      serializeInnerList({ items: ['a"b\\c'], params: new Map([['f', true], ['g', false]]) }),
      '("a\\"b\\\\c");f;g=?0',
    );
  });

  for (const [id, text] of [
    ['a key that does not start a key', '1=2'],
    ['an unterminated string', 'a="oops'],
    ['a bad escape', 'a="o\\ps"'],
    ['a byte sequence that is not base64', 'a=:not base64!:'],
    ['an unterminated byte sequence', 'a=:AAAA'],
    ['a bad boolean', 'a=?2'],
    ['an item of no supported type', 'a=%bad'],
    ['an unterminated inner list', 'a=("@method"'],
    ['parameters on a covered component', 'a=("@method";q=1)'],
    ['a missing separator inside an inner list', 'a=("@method""@path")'],
    ['a missing comma between members', 'a=1 b=2'],
    ['a trailing comma', 'a=1, '],
    ['an integer that is not one', 'a=-'],
  ]) {
    it(`refuses ${id}`, () => {
      assert.throws(() => parseDictionary(text));
    });
  }
});

describe('the defensive arms', () => {
  // Six branches that only fire when a caller omits something optional. They are cheap to reach
  // and expensive to leave untested: an untaken arm in a parser or a header walk is where a
  // "cannot happen" turns into somebody else's exception.

  it('builds a base for a request with no headers at all', () => {
    const base = signatureBase({
      method: 'GET', url: 'https://example.com/f', covered: ['@method'], created: 1, keyid: 'k',
    });
    assert.equal(new TextDecoder().decode(base).split('\n')[0], '"@method": GET');
  });

  it('signs a request with no headers at all', async () => {
    const headers = await signRequest({ key, method: 'GET', url: 'https://example.com/f', created: 1 });
    assert.ok(headers['Signature-Input'].startsWith('sig=('));
  });

  it('signs a body on a request with no headers at all, and returns the digest it made', async () => {
    const headers = await signRequest({
      key, method: 'POST', url: 'https://example.com/f', body: BODY, created: 1,
    });
    assert.ok(headers['Content-Digest'].startsWith('sha-256=:'));
    const verdict = await verifyRequest({
      method: 'POST', url: 'https://example.com/f', headers, body: BODY, maxAge: null,
    });
    assert.equal(verdict.aid, AID);
  });

  it('refuses to verify a request with no headers at all', async () => {
    await assert.rejects(
      () => verifyRequest({ method: 'GET', url: '/f', maxAge: null }),
      errors.MissingSignatureInput,
    );
  });

  it('refuses a parameter name that runs off the end of the field', () => {
    assert.throws(() => parseDictionary('a;'));
  });

  it('refuses a member whose value runs off the end of the field', () => {
    assert.throws(() => parseDictionary('a='));
  });
});
